### ISSUE TITLE
MemoryUtilization cloudwatch metric is being reported incorrectly

### ISSUE THREAD

**Issue by ryanwalls** [aws/amazon-ecs-agent#354]
Agent: 1.8.1
Docker: 1.10.3

We have 1 instance in our cluster with 15 GB of memory. We ran a task on it that is registered for 7 GB of memory. 

I've attached a newrelic graph showing the actual memory usage and the graph of our cloudwatch MemoryUtilization metric that shows the incorrect memory usage.

<img width="428" alt="cloudwatch metrics" src="https://cloud.githubusercontent.com/assets/842653/14032520/0483a232-f1d9-11e5-80a8-e2c7e375b284.png">
<img width="846" alt="newrelic metrics" src="https://cloud.githubusercontent.com/assets/842653/14032519/048347e2-f1d9-11e5-82f8-c6cfc5c1a9d5.png">

Bottom line: It appears that the cloudwatch MemoryUtilization metric for the cluster is very inaccurate. 

Note: The task was killed in ECS because it reached the 7 GB limit, even though our cloudwatch graph never showed it even approaching 7 GB.

**Comment by aaithal** [comment]
@ryanwalls Could you please clarify if the attached graph is for Cluster or Service utilization? Or, what are the dimensions of these metrics? Is it just `ClusterName` or is it `(ClusterName,ServiceName)`?

Thanks,
Anirudh

**Comment by ryanwalls** [comment]
Cluster utilization.  Dimension is just `ClusterName`.

**Comment by aaithal** [comment]
Thanks for that information. As described [here](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/cloudwatch-metrics.html#cluster_utilization), the cluster utilization shows the utilization for the entire cluster and not just the task. So, it is a function of both number of instances in the cluster and the type of instances in the cluster. Please let us know if this math adds up for the Task that you launched. If not, please provide the following information and we can determine if there's a bug on our end:
* Cluster Arn
* Container Instances' Arn in the Cluster
* Task ID (not necessary, but good to have)

You can send them me at *aithal at amazon.com*

**Comment by ryanwalls** [comment]
Sent the details.

**Comment by aaithal** [comment]
Thanks. Also, can you please verify if this corresponds to what's being reported when you run the `docker stats` command? If there's a discrepancy between that and what's being reported in Cloudwatch, we'd need to take a look at [agent debug logs](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/troubleshooting.html#logs).

**Comment by aaithal** [comment]
The discussion in #280 seems very pertinent to what you are observing, based on the behavior you describe. If you have a similar use-case or feature request, feel free to +1 as well.

**Comment by ryanwalls** [comment]
@aaithal I think that could be related, but I also feel that the `MemoryUtilization` metric does not match what `docker stats` outputs.  On the graph I sent you, when I run `docker stats`, the container reports 7 GB / 7 GB in use, whereas `MemoryUtilization` is reporting ~20% of 15 GB (so about 3 GB).

**Comment by aaithal** [comment]
@ryanwalls Thanks for providing these details. I was able to reproduce this issue on our end as well. There is a bug in the agent code, where if a container's memory utilization > 4294967295 bytes, the ECS Agent is reporting it wrongly to the backend. We are working on a fix for the same.

**Comment by ryanwalls** [comment]
@aaithal Awesome.  Thanks for the update.
